ENH: Add fixed_params to Hannan Rissanen (GH7202)

jackzyliu opened the pull request:
- [x] closes #7202
- [x] xref #6159
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message). 

<details>

**Notes**:

* It is essential that you add a test when making code changes. Tests are not 
  needed for doc changes.
* When adding a new function, test values should usually be verified in another package (e.g., R/SAS/Stata).
* When fixing a bug, you must add a test that would produce the bug in main and
  then show that it is fixed with the new code.
* New code additions must be well formatted. Changes should pass flake8. If on Linux or OSX, you can
  verify you changes are well formatted by running 
  ```
  git diff upstream/main -u -- "*.py" | flake8 --diff --isolated
  ```
  assuming `flake8` is installed. This command is also available on Windows 
  using the Windows System for Linux once `flake8` is installed in the 
  local Linux environment. While passing this test is not required, it is good practice and it help 
  improve code quality in `statsmodels`.
* Docstring additions must render correctly, including escapes and LaTeX.

</details>

jackzyliu commented on statsmodels/tsa/arima/estimators/hannan_rissanen.py:
```diff
@@ -106,23 +114,46 @@ def hannan_rissanen(endog, ar_order=0, ma_order=0, demean=True,
     _ = SARIMAXSpecification(endog, ar_order=initial_ar_order)
 
+    # Unpack fixed and free ar/ma lags, ix, and params (fixed only)
+    params_info = _package_fixed_and_free_params_info(
```
`params_info` contains `free_ar_ix`, `fixed_ar_ix`, `free_ma_ix`, `fixed_ma_ix`, etc. Please see `_package_fixed_and_free_params_info` below.

jackzyliu commented on statsmodels/tsa/arima/estimators/hannan_rissanen.py:
```diff
@@ -106,23 +114,46 @@ def hannan_rissanen(endog, ar_order=0, ma_order=0, demean=True,
-    # for indexing purposes, must have dtype int)
-    ar_ix = np.array(spec.ar_lags, dtype=int) - 1
-    lagged_endog = lagmat(endog, max_ar_order, trim='both')[:, ar_ix]
+    lagged_endog = lagmat(endog, max_ar_order, trim='both')
```
Delaying the column slicing here for easier separation between cols with free params vs. cols with fixed params later.

jackzyliu commented on statsmodels/tsa/arima/estimators/hannan_rissanen.py:
```diff
@@ -106,23 +114,46 @@ def hannan_rissanen(endog, ar_order=0, ma_order=0, demean=True,
+            res = mod.fit()
+            resid = res.resid
+            p.sigma2 = res.scale
+            p.ar_params = _stitch_fixed_and_free_params(
```
Using a simple `_stitch_fixed_and_free_params` function here to 1) combine free and fixed params, and 2) sort combined params by the order of lag. Please see `_stitch_fixed_and_free_params` below.

jackzyliu commented on statsmodels/tsa/arima/estimators/hannan_rissanen.py:
```diff
@@ -133,22 +164,57 @@ def hannan_rissanen(endog, ar_order=0, ma_order=0, demean=True,
         if unbiased is True or unbiased is None:
-            if p.is_stationary and p.is_invertible:
+            if len(fixed_params) != 0:
+                raise NotImplementedError
```
Raising `NotImplementedError` here as instructed.

jackzyliu commented on statsmodels/tsa/arima/estimators/tests/test_hannan_rissanen.py:
```diff
@@ -100,3 +106,190 @@ def test_unbiased_error():
+        {"ar.L1": 0.69607715, "ma.L1": 0.37879692},  # no free params
+    ]
+)
+def test_itsmr_with_fixed_params(fixed_params):
```
Here are a few examples of fixing 1 or more parameters for Example 5.1.7 in Brockwell and Davis (2016).

jackzyliu commented:
@ChadFulton, sorry about the delay. I was on vacation for a while. 

I just added the `fixed_params` feature within `hannan_rissanen` as well as a few corresponding tests. As an example, the following copy-and-paste-able code snippet fixes the (original) AR1 parameter for Example 5.1.7 in Brockwell and Davis (2016). It is also added as a test case in `test_itsmr_with_fixed_params`. 

```
from statsmodels.tsa.arima.estimators import hannan_rissanen
from statsmodels.tsa.arima.datasets.brockwell_davis_2002 import lake
endog = lake.copy()
p, other_results = hannan_rissanen.hannan_rissanen(
    endog, ar_order=1, ma_order=1, demean=True,
    initial_ar_order=22, unbiased=False,
    fixed_params={"ar.L1": 0.69607715}
)
```

The outputs (`p` and `other_results`) are attached below. 
<details> 

```
In[9]: p
Out[9]: SARIMAXParams(ar=[0.69607715], ma=[0.37879692], sigma2=0.501446209619244)
In[10]: other_results
Out[10]: 
{'spec': SARIMAXSpecification(endog=y, order=(1, 0, 1)),
 'initial_ar_order': 22,
 'resid': array([-0.12639734,  0.37973763, -0.45249946,  0.54390611, -0.45539779,
         0.2174877 ,  0.71559074, -0.07364959,  0.14143549,  0.44235971,
         0.3059347 , -0.30396086, -0.32157663, -0.3786135 ,  0.41014448,
         0.64865104, -0.5935746 , -0.60939023,  1.2796705 ,  0.44299881,
         0.14379167, -0.3694259 , -0.08080896, -0.52753884,  0.33941018,
        -0.8914752 , -0.20873552, -1.4495323 , -0.11090343,  0.47671633,
         0.33652349,  1.68131805, -1.23099003, -1.55226164, -0.38727042,
        -0.48729428, -1.20839297,  0.141095  , -0.75493403, -0.40589502,
         0.35554935,  0.03529988, -0.9417608 , -0.54119365,  0.7795636 ,
         0.63509188, -0.61454623,  0.55152717, -0.15665677,  0.1978954 ,
        -0.37319131, -1.038126  ,  0.07622795,  1.31210166,  1.05728195,
        -0.03461711, -0.02664505, -0.04312429, -0.70688682, -0.441793  ,
        -1.08417358, -0.36757403,  1.46062506, -1.29794955, -0.00809655,
        -1.40204363, -1.24171812,  0.28769771,  0.03943561,  0.17817281,
        -0.1824329 ,  1.08071756, -0.50266338,  0.79617474, -0.1034897 ])}
```
</details>

I tried to keep the original output structure completely unchanged (i.e. fixed params are not omitted in `p`) for 2 reasons:
1. It should be conveniently consistent with the outputted 'spec'
2. I think `ARIMA.fit` builds results for the few methods (including `hannan_rissanen`) assuming fixed parameters are present (referencing [here](https://github.com/statsmodels/statsmodels/blob/main/statsmodels/tsa/arima/model.py#L449-L450), specifically `includes_fixed=True`).

Please let me know if we should omit the fixed params in output. 

Additionally, I have a few **questions** that I would like to hear your thought on:
1. The code does not apply fixed params to Yule-Walker in the first step of Hannan-Rissanen, but maybe we should? It's not intuitive to me whether/how much that changes result optimality.
2. I added a few helpers that may be re-used for adding fixed params for `innovations` and `gls` (e.g. `_validate_fixed_params` and `_package_fixed_and_free_params_info`). Maybe we should fold these under `SARIMAXSpecification`?
3. Which version file should I add the release note to?

Once this PR is done. I will do another PR to enable Hannan-Rissanen with fixed params in `ARIMA.fit`.

ChadFulton commented on statsmodels/tsa/arima/estimators/tests/test_hannan_rissanen.py:
```diff
@@ -100,3 +106,190 @@ def test_unbiased_error():
+        )
+    else:
+        valid_params = sorted(list(set(spec.param_names) - {'sigma2'}))
+        msg = f"Invalid fixed parameter(s): {invalid_fixed_params}. " \
```
Style-wise, we typically use parentheses instead of the `\` operator.

ChadFulton reviewed (commented):
This is great, thanks very much for this PR!  I just noted two minor things, and after that it should be ready to merge.

ChadFulton commented:
> I tried to keep the original output structure completely unchanged (i.e. fixed params are not omitted in p) for 2 reasons:
> 
> 1. It should be conveniently consistent with the outputted 'spec'
> 2. I think ARIMA.fit builds results for the few methods (including hannan_rissanen) assuming fixed parameters are present (referencing here, specifically includes_fixed=True).

Yes, I agree that keeping the fixed parameters in the parameter vector is the right thing to do.

Once we start incorporating these into e.g. the `GLS` setup or the `ARIMA` model setup, then we probably will need to find a way to indicate which parameters were fixed and which were freely estimated. At a first glance, it seems like this might be best to do in `SARIMAXParams`; what do you think?

> Additionally, I have a few questions that I would like to hear your thought on:
> 
> 1. The code does not apply fixed params to Yule-Walker in the first step of Hannan-Rissanen, but maybe we should? It's not intuitive to me whether/how much that changes result optimality.

I think that it is okay to do what you've done here. This first step is just to get estimates of the residuals (and e.g. the AR lag order is not the same as in the specified model).

> 2. I added a few helpers that may be re-used for adding fixed params for innovations and gls (e.g. _validate_fixed_params and _package_fixed_and_free_params_info). Maybe we should fold these under SARIMAXSpecification?

I agree that `SARIMAXSpecification` is a natural place for them to be, but I don't think you need to move them yet unless you feel like it (since they're private methods, we can always move them later).

> 3. Which version file should I add the release note to?

This feature should be part of v0.13, but we do not yet have a file for release notes.  Please feel free to create one at `docs/source/release/version0.13.rst`, you can just do a stub like the following:

```rst
:orphan:

==============
Release 0.12.0
==============

Release summary
===============

statsmodels is using github to store the updated documentation. Two version are available:

- `Stable <https://www.statsmodels.org/>`_, the latest release
- `Development <https://www.statsmodels.org/devel/>`_, the latest build of the main branch

**Warning**

API stability is not guaranteed for new features, although even in
this case changes will be made in a backwards compatible way if
possible. The stability of a new feature depends on how much time it
was already in statsmodels main and how much usage it has already
seen.  If there are specific known problems or limitations, then they
are mentioned in the docstrings.

Stats
-----
**Issues Closed**: TBD

**Pull Requests Merged**: TBD

The Highlights
==============

Time series analysis
--------------------

Fixed parameters in ARIMA estimators
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

<your text here>
```

> Please let me know of any comments or questions. Once this PR is done. I will do another PR to enable Hannan-Rissanen with fixed params in ARIMA.fit.

That's great, looking forward to it!

jackzyliu commented on statsmodels/tsa/arima/estimators/hannan_rissanen.py:
```diff
@@ -133,22 +164,57 @@ def hannan_rissanen(endog, ar_order=0, ma_order=0, demean=True,
+            resid = res.resid
+            p.sigma2 = res.scale
 
         # Step 3: bias correction (if requested)
```
Ah, this definitely makes more sense. I was actually wondering what `unbiased=None` with fixed params should mean.

jackzyliu commented on statsmodels/tsa/arima/estimators/tests/test_hannan_rissanen.py:
```diff
@@ -100,3 +106,238 @@ def test_unbiased_error():
         hannan_rissanen(endog, ar_order=1, ma_order=1, unbiased=True)
+
+
+def test_set_default_unbiased():
```
Adding this test and `test_set_default_unbiased_with_fixed_params` below to test the added handling of `unbiased` and `fixed_params` args.

jackzyliu commented on statsmodels/tsa/arima/estimators/hannan_rissanen.py:
```diff
@@ -133,69 +164,125 @@ def hannan_rissanen(endog, ar_order=0, ma_order=0, demean=True,
+
+        # Step 3.1: validate `unbiased` argument and handle setting the default
+        if unbiased is True:
+            if len(fixed_params) != 0:
```
Checking `fixed_params` first here to avoid unnecessary `p.is_stationary` and `p_is_invertible` computations.

jackzyliu commented:
@ChadFulton Thanks a lot for the review and help! 

> Once we start incorporating these into e.g. the GLS setup or the ARIMA model setup, then we probably will need to find a way to indicate which parameters were fixed and which were freely estimated. At a first glance, it seems like this might be best to do in SARIMAXParams; what do you think?

This makes a lot of sense. The code change should be pretty non-invasive too, which I think just requires adding a few new pairs of setters and getters like `.is_ar_param_fixed`, `.is_ma_param_fixed`, `.is_param_fixed`, `.free_ar_params`, `.fixed_ma_params`, etc. that are compatible with the existing class. Then, before we extend `SARIMAXSpecification` to incorporate fixed parameters, we can simply: 
1. first initialize SARIMAXParams like before;
2. then use `.is_ar_param_fixed` and `.is_ma_param_fixed` to indicate which params are fixed
3. lastly, either use `.ar_params` and `.ma_params` like before or call those added free and fixed param setters separately to set the params.

This should also be directly backward-compatible. 

> This feature should be part of v0.13, but we do not yet have a file for release notes. Please feel free to create one at docs/source/release/version0.13.rst, you can just do a stub like the following:

I created a `version0.13.rst` from your template and added a few more sections by referencing `versions0.12.rst`. Please let me know if anything needs changing.

ChadFulton commented:
This all looks good to me.  I'll merge tomorrow morning unless there's anything else you want to do first. Thanks again!

> This makes a lot of sense. The code change should be pretty non-invasive too, which I think just requires adding a few new pairs of setters and getters like .is_ar_param_fixed, .is_ma_param_fixed, .is_param_fixed, .free_ar_params, .fixed_ma_params, etc. that are compatible with the existing class. 

I agree, this all makes sense to me.

jackzyliu commented:
Thanks! I will get on enable Hannan-Rissanen with fixed params in `ARIMA.fit` right afterwards.

Maxnoue reviewed (commented):
Help me pls

bashtage commented:
@Maxnoue help with what?
